File: backend/app/routers/solve.py
from fastapi import APIRouter, Depends, HTTPException, status

from app.core.config import get_db
from app.core.auth import get_current_user
from app.models.solve import SolveModel
from app.schemas.solve import SolveRequest, SolveSchema, StatsSchema
from app.services.solve_service import SolveService

from datetime import datetime

router = APIRouter()

# Get all solves by user's id
@router.get("/solve", response_model=list[SolveSchema])
def get_solve(
    current_user: dict = Depends(get_current_user),  # ← Add this parameter
    db = Depends(get_db)
):
    user_id = current_user["user_id"]

    solves = db.query(SolveModel).filter(SolveModel.user_id == user_id).all()

    print("solves:", solves)
    return solves

@router.post("/solve", response_model=SolveSchema)
def create_solve(
    req: SolveRequest,
    current_user: dict = Depends(get_current_user),
    db = Depends(get_db)
):
    user_id = current_user["user_id"]

    # Create new solve object
    solve = SolveModel(
        user_id=user_id,
        time=req.time,
        cube_type=req.cube_type,
        scramble=req.scramble,
        dnf=req.dnf,
        createdAt=datetime.now(),
    )

    # Adding solve to database
    db.add(solve)
    db.commit()
    db.refresh(solve)

    return solve

@router.get("/stats", response_model=StatsSchema)
def get_stats(
    current_user: dict = Depends(get_current_user),
    db = Depends(get_db)
):
    user_id = current_user["user_id"]

    # getting all of the solves
    solves = db.query(SolveModel).filter(SolveModel.user_id == user_id).all()

    stat_calculator = SolveService()

    times = [solve.time for solve in solves]

    stats = {
        "best_ao5": stat_calculator.calculate_best_ao5(times),
        "best_ao12": stat_calculator.calculate_best_ao12(times),
        "best_ao100": stat_calculator.calculate_best_ao100(times),
        "best_time": stat_calculator.get_best_time(times),
        "total_solves": len(solves)
    }

    return stats

@router.delete("/solve/{solve_id}")
def delete_solve(
    solve_id: int,
    current_user: dict = Depends(get_current_user),
    db = Depends(get_db)
):
    user_id = current_user["user_id"]
    
    solve = db.query(SolveModel).filter(
        SolveModel.id == solve_id, 
        SolveModel.user_id == user_id
    ).first()
    
    if not solve:
        raise HTTPException(
            status_code=404,
            detail="Solve not found"
        )
    
    db.delete(solve)
    db.commit()
    
    return {"status": "success"}

File: convex/solve.ts
import { mutation, query } from "./_generated/server";
import { v } from "convex/values";
import { verifyToken } from "./auth";

export const getSolves = query({
  args: {
    token: v.string(),
  },
  handler: async (ctx, args) => {
    const decoded = await verifyToken(args.token);
    if (!decoded) {
      throw new Error("Invalid token");
    }

    return await ctx.db
      .query("solves")
      .filter((q) => q.eq(q.field("userId"), decoded.userId))
      .collect();
  },
});

export const createSolve = mutation({
  args: {
    token: v.string(),
    cubeType: v.string(),
    time: v.number(),
    scramble: v.string(),
    dnf: v.boolean(),
  },
  handler: async (ctx, args) => {
    const decoded = await verifyToken(args.token);
    if (!decoded) {
      throw new Error("Invalid token");
    }

    return await ctx.db.insert("solves", {
      userId: decoded.userId,
      cubeType: args.cubeType,
      time: args.time,
      scramble: args.scramble,
      dnf: args.dnf,
    });
  },
});

export const deleteSolve = mutation({
  args: {
    token: v.string(),
    solveId: v.id("solves"),
  },
  handler: async (ctx, args) => {
    const decoded = await verifyToken(args.token);
    if (!decoded) {
      throw new Error("Invalid token");
    }

    const solve = await ctx.db.get(args.solveId);
    if (!solve || solve.userId !== decoded.userId) {
      throw new Error("Solve not found or access denied");
    }

    await ctx.db.delete(args.solveId);
    return { status: "success" };
  },
});

export const getStats = query({
  args: {
    token: v.string(),
  },
  handler: async (ctx, args) => {
    const decoded = await verifyToken(args.token);
    if (!decoded) {
      throw new Error("Invalid token");
    }

    const solves = await ctx.db
      .query("solves")
      .filter((q) => q.eq(q.field("userId"), decoded.userId))
      .collect();

    const times = solves.map((s) => s.time).filter((t) => !isNaN(t));

    const calculateAO = (arr: number[], n: number) => {
      if (arr.length < n) return 0;
      let best = Infinity;
      for (let i = 0; i <= arr.length - n; i++) {
        const avg =
          arr.slice(i, i + n).reduce((a, b) => a + b, 0) / n;
        if (avg < best) best = avg;
      }
      return best === Infinity ? 0 : parseFloat(best.toFixed(2));
    };

    return {
      best_ao5: calculateAO(times, 5),
      best_ao12: calculateAO(times, 12),
      best_ao100: calculateAO(times, 100),
      best_time: times.length > 0 ? Math.min(...times) : 0,
      total_solves: solves.length,
    };
  },
});
